File: ios/Cast/CastReSign.swift
import Foundation
import os.log

#if AUDIOBROWSER_ENABLE_CAST

  import GoogleCast

  /// Bounded, reactive re-signing of stale media URLs on the Cast receiver.
  ///
  /// ADR-0003: the full Queue is mirrored onto the receiver and each item's
  /// media URL is signed at load time. Over multi-hour live sessions a signed
  /// URL deep in the queue expires before the receiver reaches it. When the
  /// receiver hits a **stale-URL load error**, we JIT re-resolve *that one item*
  /// (`BrowserManager.resolveMediaUrl(target: .cast)`) and push it back with
  /// `GCKRemoteMediaClient.queueUpdateItems`.
  ///
  /// The attempt count is **capped per item** so a genuinely dead stream can't
  /// drive an infinite reload loop (mirrors the `StuckRecoveryPolicy`
  /// philosophy): once an item exhausts its budget we stop re-signing and let
  /// the error surface.
  ///
  /// Distinguishing "stale URL" from "dead stream" is best-effort and is an
  /// on-device verification risk noted in TODO-CAST.md (open risk #2): we treat
  /// the first few media load failures on an item as recoverable, then give up.
  @MainActor
  final class CastReSign {
    private let logger = Logger(subsystem: "com.audiobrowser", category: "CastReSign")

    /// Re-resolves a track's media URL for the Cast device. Injected so this
    /// type stays decoupled from BrowserManager (set by CastSessionManager to
    /// `browserManager.resolveMediaUrl(_, track:, target: .cast)`).
    var resolveCastUrl: ((_ src: String, _ track: Track) async -> String)?

    /// Pure per-item attempt cap + in-flight dedup (testable, no SDK). This class
    /// keeps only the SDK glue (read identity, resolve, `queueUpdate`).
    private var budget: CastReSignBudget

    init(maxAttemptsPerItem: Int = 3) {
      budget = CastReSignBudget(maxAttemptsPerItem: maxAttemptsPerItem)
    }

    /// Clear all attempt accounting (call on a fresh queueLoad).
    func reset() {
      budget.reset()
    }

    /// React to a remote media load error. Returns true if a re-sign was
    /// dispatched OR one is already in flight for this item (the caller should NOT
    /// surface a terminal error yet); false if the item is out of budget /
    /// unrecoverable (the caller surfaces the real error).
    ///
    /// - Parameters:
    ///   - remoteClient: the active `GCKRemoteMediaClient`.
    ///   - itemID: the receiver queue item that failed (from
    ///     `remoteMediaClient.mediaStatus?.currentItemID` or the error context).
    @discardableResult
    func handleLoadError(remoteClient: GCKRemoteMediaClient, itemID: UInt) -> Bool {
      // A re-sign for this item is already dispatched — wait for it to land. Checked
      // BEFORE the identity read so a transient gap in the receiver's queue (the item
      // momentarily absent mid-resign) can't flip an in-flight item to a false
      // terminal error.
      if budget.isInFlight(itemID) { return true }

      guard let resolveCastUrl else {
        logger.error("CastReSign: no resolver wired — cannot re-sign item \(itemID)")
        return false
      }

      // Read the current item's stashed Track identity off the receiver BEFORE
      // claiming a budget attempt — a missing identity is unrecoverable, but it must
      // not consume one of the few re-sign attempts.
      guard let item = remoteClient.mediaQueue.item(withID: itemID, fetchIfNeeded: false) ?? remoteClient.mediaStatus?.queueItem(withItemID: itemID),
            let track = CastMediaItemConverter.track(from: item.mediaInformation),
            let src = track.src
      else {
        logger.error("CastReSign: could not read Track identity for item \(itemID)")
        return false
      }

      switch budget.shouldAttempt(itemID) {
      case .inFlight:
        // A re-sign is already running for this item — wait for it to land.
        return true
      case .exhausted:
        logger.error("CastReSign: item \(itemID) exhausted re-sign budget — surfacing error")
        return false
      case .attempt:
        break
      }

      logger.info("CastReSign: re-signing item \(itemID)")
      Task { @MainActor in
        // Clear the in-flight mark once this attempt settles (success or not), so
        // a later genuine expiry of the same item can be re-signed again.
        defer { self.budget.markDone(itemID) }
        let freshUrl = await resolveCastUrl(src, track)
        let artworkUrl = track.artwork
        guard let updated = CastMediaItemConverter.queueItem(
          for: track, mediaUrl: freshUrl, artworkUrl: artworkUrl, autoplay: true,
        ) else {
          self.logger.error("CastReSign: re-resolved URL for item \(itemID) is unparseable — giving up")
          return
        }
        // queueUpdateItems replaces the item in place on the receiver without a
        // full reload (which would drop live audio).
        remoteClient.queueUpdate([updated], byReorderingWithIDs: nil, insertBeforeItemWithID: kGCKMediaQueueInvalidItemID)
      }
      return true
    }
  }

#endif

File: android/src/cast/java/com/audiobrowser/cast/CastReSign.kt
package com.audiobrowser.cast

import com.audiobrowser.browser.BrowserManager
import com.audiobrowser.browser.resolveMediaUrl
import com.google.android.gms.cast.MediaInfo
import com.google.android.gms.cast.MediaQueueItem
import com.google.android.gms.cast.framework.media.RemoteMediaClient
import com.margelo.nitro.audiobrowser.MediaResolveTarget
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.json.JSONObject
import timber.log.Timber

/**
 * Bounded, reactive re-signing of stale Cast queue URLs (see ADR 0003). Signed stream URLs
 * expire over multi-hour live sessions, so an item deep in the mirrored queue can be dead by the
 * time the receiver reaches it. When the receiver hits a load error, we JIT re-resolve *that one
 * item* via [BrowserManager.resolveMediaUrl] with `target=cast` and update it on the receiver with
 * [RemoteMediaClient.queueUpdateItems].
 *
 * Attempts are capped per item id (mirroring `StuckRecoveryPolicy`'s philosophy) so a genuinely
 * dead stream surfaces a real error instead of looping forever.
 */
class CastReSign(
  private val browserManager: () -> BrowserManager?,
  private val scope: CoroutineScope,
  maxAttemptsPerItem: Int = 3,
) {
  // Pure per-item cap + in-flight dedup bookkeeping (testable, in main). This class keeps only the
  // SDK glue (resolve + queueUpdateItems).
  private val budget = CastReSignBudget(maxAttemptsPerItem)

  /** Resets the per-item attempt budget — call on a fresh queue load / new Cast session. */
  fun reset() {
    budget.reset()
  }

  /**
   * Attempts a re-sign of the receiver's currently-loading item. Returns true if a re-sign was
   * dispatched OR one is already in flight for this item (caller should NOT surface a terminal
   * error yet), false if the budget is exhausted or the item can't be re-resolved (caller surfaces
   * the real error). Call on the main thread (RemoteMediaClient callback thread) — the budget is
   * not synchronized.
   */
  fun onLoadError(remoteMediaClient: RemoteMediaClient): Boolean {
    val item = remoteMediaClient.currentItem ?: return false
    val track = CastTrackCodec.fromCustomData(item.media?.customData) ?: return false
    // Re-resolution needs a raw src — bail (do NOT touch the budget) when there isn't one.
    val originalUrl = track.src ?: return false
    // Capture the failing item id now, at detection — not inside the coroutine, where currentItem
    // may have moved on.
    val key = track.src ?: track.url ?: track.id ?: return false

    when (budget.shouldAttempt(key)) {
      CastReSignBudget.Decision.IN_FLIGHT -> return true
      CastReSignBudget.Decision.EXHAUSTED -> {
        Timber.w("Cast re-sign budget exhausted for item key=$key — surfacing error")
        return false
      }
      CastReSignBudget.Decision.ATTEMPT -> Unit
    }
    val manager = browserManager() ?: return false

    scope.launch {
      try {
        val resolved = manager.resolveMediaUrl(originalUrl, MediaResolveTarget.CAST)
        val freshUrl = resolved?.path ?: originalUrl
        val updated = reSignedQueueItem(item, freshUrl)
        remoteMediaClient.queueUpdateItems(arrayOf(updated), JSONObject())
        Timber.d("Cast re-sign dispatched for item key=$key")
      } catch (e: Exception) {
        Timber.e(e, "Cast re-sign failed for item key=$key")
      } finally {
        budget.markDone(key)
      }
    }
    return true
  }

  /** Rebuilds a queue item with a fresh content id (the re-resolved, self-contained URL). */
  private fun reSignedQueueItem(item: MediaQueueItem, freshUrl: String): MediaQueueItem {
    val oldInfo = item.media
    val newInfo =
      MediaInfo.Builder(freshUrl)
        // Preserve the original stream type — for a live stream this stays STREAM_TYPE_LIVE so the
        // re-signed item isn't downgraded to a finite buffered file.
        .setStreamType(oldInfo?.streamType ?: MediaInfo.STREAM_TYPE_BUFFERED)
        // Concrete MIME from the fresh URL (shared helper); never the bogus "audio/*".
        .setContentType(oldInfo?.contentType ?: CastMediaItemConverter.contentTypeFor(freshUrl))
        .setMetadata(oldInfo?.metadata)
        .setCustomData(oldInfo?.customData)
        .build()
    // MediaQueueItem.Builder has no setMedia(): rebuild from the fresh MediaInfo and carry over the
    // item-level fields. Skip startTime when unset (NaN) — Builder.setStartTime(NaN) throws.
    return MediaQueueItem.Builder(newInfo)
      .setAutoplay(item.autoplay)
      .setPreloadTime(item.preloadTime)
      .apply {
        item.customData?.let { setCustomData(it) }
        if (!item.startTime.isNaN()) setStartTime(item.startTime)
      }
      .build()
  }
}
